File: algorithms/cpp/distributeCoinsinBinaryTree/distributeCoinsinBinaryTree.cpp
// Source: https://leetcode.com/problems/distribute-coins-in-binary-tree/
// Author: Miao Zhang
// Date:   2021-03-30

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int distributeCoins(TreeNode* root) {
        int res = 0;
        balance(root, res);
        return res;
    }

private:
    int balance(TreeNode* root, int& res) {
        if (!root) return 0;
        int l = balance(root->left, res);
        int r = balance(root->right, res);
        res += abs(l) + abs(r);
        return l + r + root->val - 1;
    }
};

File: algorithms/python/distributeCoinsinBinaryTree/distributeCoinsinBinaryTree.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/distribute-coins-in-binary-tree/
# Author: Miao Zhang
# Date:   2021-03-30

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: TreeNode) -> int:
        self.res = 0
        
        def dfs(root):
            if not root: return 0
            l = dfs(root.left)
            r = dfs(root.right)
            self.res += abs(l) + abs(r)
            return l + r + root.val - 1
        
        dfs(root)
        return self.res
